Feat: spinner on Check Status button

thekaveman opened the pull request:
Closes #948

Rebased @machikoyasuda's work from #1043 on the new reCAPTCHA fixes. 

Also fixed the ordering of the jQuery function calls to ensure the `disabled` state is applied to the _button_ and not the _text_ element inside.

## Demo

![spinner](https://user-images.githubusercontent.com/1783439/198087145-b2361de7-7901-4165-8e20-f34b5fd68bb7.gif)

machikoyasuda commented on benefits/core/templates/core/includes/form.html:
```diff
@@ -47,10 +50,11 @@
             $("button[type=submit]", this)
+              .removeClass("spinner-hidden")
+              .addClass("disabled")
               .attr("role", "status")
```
I wonder if this is necessary if `<span class="spinner-border spinner-border-sm" role="status"></span>` already has `role="status"`? I was not sure.

machikoyasuda replied:
After reading this https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Roles/status_role, I don't think we should have two elements with `role=status` b/c it would announce the same status twice.

thekaveman replied:
Fixed in [`b0343d6`](https://github.com/cal-itp/benefits/pull/1088/commits/b0343d6b1f37ce26b5f107ae531219f16dd3d938)

angela-tran reviewed (commented):
This looks good to me. I think @machikoyasuda is right about not needing to call `.attr("role", "status")`.

machikoyasuda reviewed (commented):
I'm testing this locally on my machine with reCaptcha enabled:
- Clicking `Check Status` without filling out two fields properly gives me the browser validation to `Please fill out this field` on the first field. Button does not animate.
- Clicking `Check Status` with a badly formatted first field properly gives me the browser validation to `Please match the requested format`.  Button does not animate.
- Clicking `Check Status` with a properly formatted first field and empty second field proerly gives me the validation to `Please fill out this field` on the second field. Button does not animate.
- Clicking `Check Status` after filling out the last name field at _this_ point, after having done all these bad interactions, I'm not getting the Button animation.

Not sure if it's my local machine/reCaptcha being wonky or not.

machikoyasuda reviewed (changes requested):
Remove the `.attr("role", "status")`.

And make a new issue/PR to: Remove the X from this error bar

<img width="1432" alt="image" src="https://user-images.githubusercontent.com/3673236/198110577-67613881-4387-4e5d-982f-88acaefc41d2.png">
<img width="653" alt="image" src="https://user-images.githubusercontent.com/3673236/198110650-5b063e26-491c-41c9-9a0d-70c7fbeb1eea.png">


After merging this PR, we need to extensively test all the listed above form interactions for the CC form specifically on both desktop and mobile with recaptcha properly enabled.

angela-tran commented:
> ...
> - Clicking `Check Status` after filling out the last name field at _this_ point, after having done all these bad interactions, I'm not getting the Button animation.
> - I clicked `Check Status` _with_ my dev tools open.....and the button animated 🙃 
> 
> Not sure if it's my local machine/reCaptcha being wonky or not. 

Good catch @machikoyasuda. I see the same behavior on my machine

I think this is because on a successful reCAPTCHA response, we're [calling `submit()`](https://github.com/cal-itp/benefits/blob/e987616d06b1dfb20b254deeb06e1ac135023c5f/benefits/core/templates/core/includes/form.html#L83) which [does not fire a `submit` event](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/submit). The spinner is [listening](https://github.com/cal-itp/benefits/pull/1088/files#diff-e80e374594a3ed9b08de85983db18bdb5fb593fcab674d7a45c57513a0a10088R50) for a `submit` event.

Maybe we can try calling [`requestSubmit()`](https://developer.mozilla.org/en-US/docs/Web/API/HTMLFormElement/requestSubmit) instead?

angela-tran commented:
Tried out calling `requestSubmit()` instead, and it fixes the issue.

machikoyasuda commented:
On further thought, I think the `role=status` from `.spinner-border` should be removed, and the `.attr("role", "status")` should remain. My thought process on this is that the role=status should be read / alerted when the spinner turns on, not when it's hidden.

thekaveman commented:
Ah, I think this is my fault in the rebase. 

We should have a `form="{{ form.id }}"` attribute on the button, which I added in #1071

thekaveman commented:
> On further thought, I think the `role=status` from `.spinner-border` should be removed, and the `.attr("role", "status")` should remain. My thought process on this is that the role=status should be read / alerted when the spinner turns on, not when it's hidden.

This is the correct interpretation. `role=status` on the button with `disabled` is telling assistive technologies that the button is no longer an interactive element, it is now providing information/status.

thekaveman commented:
> Ah, I think this is my fault in the rebase.
> 
> We should have a `form="{{ form.id }}"` attribute on the button, which I added in #1071

Ah so this is interesting... the form attribute was definitely part of it. But the new `<span>`s we added to get the spinner and text spaced correctly are messing up the JS that reCAPTCHA is using.

This line is the problem:

```js
function recaptchaSubmit($event) {
  // Checks the validity of the form. Return if invalid; HTML5 validation errors should display
  if (!$event.target.form.checkValidity()) { //<----- problem
    return;
  }
```

Because of the spinner, `$event.target` is no longer the `<button>`, it is the `<span>`!

Working on this...

thekaveman commented:
Thanks for your great reviews/comments @angela-tran @machikoyasuda!

Implemented a (hopefully) straightforward fix for reCAPTCHA + spinner:

* apply the `form="{{ form.id }}"` attribute to the button (I lost this in the rebase)
* change reCAPTCHA JS from using `$event.target` (the element that was clicked, now `<span class="btn-text">`) to `$event.currentTarget`, which is the element that was listening for the event that called the handler (i.e. the `<button>`): https://developer.mozilla.org/en-US/docs/Web/API/Event/Comparison_of_Event_Targets
* make the `disabled`/`status`/spinner stuff happen on a custom event `submitting` that is triggered:
  * by the normal form submit (when reCAPTCHA not enabled)
  * by the reCAPTCHA JS, just before `form.submit()`

angela-tran reviewed (dismissed):
Just one non-blocking question

angela-tran commented on benefits/core/templates/core/includes/form.html:
```diff
@@ -68,19 +79,23 @@
-                $event.target.form.elements["{{ request.recaptcha.data_field }}"].value = token;
+                // add the token to hidden form element
+                $event.currentTarget.form.elements["{{ request.recaptcha.data_field }}"].value = token;
+                // trigger the custom "submitting" event
+                // since calling .submit() does not trigger the "submit" event
+                // hence the normal handler won't apply
+                $($event.currentTarget.form).trigger("submitting");
```
This solution works for me. Just curious: what was the reasoning for triggering a custom event instead of calling `requestSubmit()` on line 98 to trigger a `submit` event?

thekaveman replied:
Sorry, I missed your suggestion to use that! Been a little scattered this afternoon... I like this better. Fixing.

thekaveman replied:
Rebased to drop the old implementation and added [`cf7f670`](https://github.com/cal-itp/benefits/pull/1088/commits/cf7f670df923ef0b8221bf34841d1a1391c9f180)

angela-tran reviewed (dismissed):
One non-blocking question

machikoyasuda commented:
Amazing gif!!!!!!

angela-tran reviewed (approved):
Nice!
